server/src/resources/user/grpc.rs

```diff
@@ -0,0 +1,258 @@
+use std::collections::HashMap;
+use tonic::{Code, Request, Response, Status};
+
+///Implementation of gRPC endpoints
+#[derive(Debug, Default, Copy, Clone)]
+pub struct UserImpl {}
+
+#[tonic::async_trait]
+impl UserRpc for UserImpl {
+    /// Takes an UUID string (user_id) to get the matching database record.
+    ///
+    /// # Example
+    /// ```
+    /// use svc_storage_client_grpc::client::user_rpc_client::UserRpcClient;
+    /// use svc_storage_client_grpc::client::Id;
+    ///
+    /// let mut user_client = UserRpcClient::connect(grpc_endpoint.clone()).await.unwrap();
+    /// let result = match user_client.user_by_id(tonic::Request::new(Id {
+    ///     id: "54acfe06-dd9b-42e8-8cb4-12a2fb2fa693"
+    /// }))
+    /// .await
+    /// {
+    ///    Ok(user) => user.into_inner(),
+    ///    Err(e) => panic!("Something went wrong retrieving the user: {}", e),
+    /// };
+    /// ```
+    async fn user_by_id(&self, request: Request<Id>) -> Result<Response<User>, Status> {
+        let id = request.into_inner().id;
+        if let Some(user) = USERS.lock().await.get(&id) {
+            memdb_info!("Found entry for User. uuid: {}", id);
+            Ok(Response::new(user.clone()))
+        } else {
+            let pool = get_psql_pool();
+            let data = UserPsql::new(&pool, Uuid::try_parse(&id).unwrap()).await;
+            match data {
+                Ok(user) => Ok(Response::new(User {
+                    id,
+                    data: Some(user.into()),
+                })),
+                Err(_) => Err(Status::not_found("Not found")),
+            }
+        }
+    }
+
+    /// Takes a `SearchFilter` object to search the database with the provided values.
+    ///
+    /// This method supports paged results. When the `search_field` and `search_value` are empty, no filters will be applied.
+    ///
+    /// # Example
+    /// ```
+    /// use svc_storage_client_grpc::client::user_rpc_client::UserRpcClient;
+    /// use svc_storage_client_grpc::client::SearchFilter;
+    ///
+    /// let mut user_client = UserRpcClient::connect(grpc_endpoint.clone()).await.unwrap();
+    /// let result = match user_client.users(tonic::Request::new(SearchFilter {
+    ///     search_field: "auth_method",
+    ///     search_value: "1",
+    ///     page_number: 1,
+    ///     results_per_page: 20
+    /// }))
+    /// .await
+    /// {
+    ///    Ok(users) => user.into_inner(),
+    ///    Err(e) => panic!("Something went wrong creating retrieving users: {}", e),
+    /// };
+    /// ```
+    async fn users(&self, request: Request<SearchFilter>) -> Result<Response<Users>, Status> {
+        let filter = request.into_inner();
+        let mut filter_hash = HashMap::<String, String>::new();
+        filter_hash.insert("column".to_string(), filter.search_field);
+        filter_hash.insert("value".to_string(), filter.search_value);
+
+        let pool = get_psql_pool();
+        match super::psql::search(&pool, &filter_hash).await {
```

Comment:
RE: error handling:
I can see few `unwrap`s which seem unhandled.
and than a lot of `match` clauses which can be easily replaced by `?`. But that's my preference to remove 4 lines of code for 1 or 2. Also I don't like the match as it pushes code to the right and if you have many nested errors to handle, it will become a christmas tree. :)

E.g.
```
        match super::psql::search(&pool, &filter_hash).await {
            Ok(users) => Ok(Response::new(users.into())),
            Err(e) => Err(Status::new(Code::Internal, e.to_string())),
        }
```
can become
`super::psql::search(&pool, &filter_hash).await?` (If error is different than returned error, then `.map_err` would need to be used)

Reply:
I'm actually not sure if the example can be rewritten that easily. How would you handle the return type change for the `Ok` match then? Both the `Ok` and the `Err` results need to be transformed into different results for GRCP (Response and Status)

Reply:
Ah, yes! Makes sense, I'll fix it.

Reply:
I don't think it needs for demo or R1, we can discuss what's better code style for R2+. I posted a services forum about this to discuss :)